### PR TITLE
2 endpoints swb

### PR DESCRIPTION
**Pull request by Scott-Borecki**
### What I Did:
- Create endpoints from database resources:
  - `GET /users/:id`
  - `GET   /users/:user_id/events`
  - `GET   /users/:user_id/gyms`
  - `GET   /users/:user_id/friendships`
  - `POST /users`
  - `PATCH /users/:id`
- Remove `first_name`, `last_name`, and `google_token` columns from `Users` table
- Create serializers for each table
- Create factory methods for testing
- Update `Users` relationships

### Road Blocks
- Adding self-referential relationships to UserSerializer.  This is my first time using the `json-apiserializer` gem, so didn't get to dig into the docs yet.

### Test Coverage
- [X] I have tested this code with RSpec
- The serializers were created with the `json-apiserializer` gem; however, we should consider adding tests for the serializers.
- Validations need to be added to endpoint request edge cases and tested.

### Questions/Notes:
- None

### CODE REVIEW

File: app/controllers/api/v1/users_controller.rb [hunk 1]
```diff
@@ -5,4 +5,33 @@ def show
...
+    render json: UserSerializer.new(user).serializable_hash, status: :created
+  end
+
+  private
+
+  # TODO: Do we want to allow them to update the info that is collected from
+  #       Google Oauth?
+  #       :full_name, :email, :google_id, :google_image_url
+  def user_params
+    params.permit(:zip_code, :summary, :goal, :availability_morning,
```
Comment (carolinectan) [inline app/controllers/api/v1/users_controller.rb:28]: Depending on how the info get passed from FE to BE, I wonder if `.require()` will need to be added?

File: app/controllers/api/v1/users_controller.rb [hunk 2]
```diff
@@ -5,4 +5,33 @@ def show
...
+                  :availability_afternoon, :availability_evening)
+  end
+
+  def user_create_params
+    params.permit(:zip_code, :summary, :goal, :availability_morning,
```
Comment (carolinectan) [inline app/controllers/api/v1/users_controller.rb:33]: Depending on how the info get passed from FE to BE, I wonder if `.require()` will need to be added?

File: app/controllers/api/v1/users_controller.rb [hunk 3]
```diff
@@ -5,4 +5,33 @@ def show
...
+    user = User.create!(user_create_params)
+
+    render json: UserSerializer.new(user).serializable_hash, status: :created
+  end
+
+  private
+
+  # TODO: Do we want to allow them to update the info that is collected from
+  #       Google Oauth?
+  #       :full_name, :email, :google_id, :google_image_url
```
Comment (Scott-Borecki) [inline app/controllers/api/v1/users_controller.rb:26]: Do we want to allow them to update the info that is collected from Google Oauth?
`:full_name`, `:email`, `:google_id`, `:google_image_url`

File: spec/factories/users.rb [hunk 4]
```diff
@@ -2,16 +2,71 @@
   factory :user do
     email { Faker::Internet.email }
     google_id { Faker::Alphanumeric.alphanumeric(number: 10) } # Update with same number of characters as id
```
Comment (Scott-Borecki) [inline spec/factories/users.rb:4]: Do we know how many characters the `google_id` is ... if it is a fixed number of characters?

File: config/routes.rb [hunk 5]
```diff
@@ -3,7 +3,11 @@
 
   namespace :api do
     namespace :v1 do
-      resources :users, only: [:show]
+      resources :users, only: [:show, :update, :create] do
+        resources :events, only: [:index], controller: 'users/events'
```
Comment (carolinectan) [inline config/routes.rb:7]: Good use of `controller:` to specify which controller. Good use of specifying `'users/events'` to use a namespaced controller.

File: db/migrate/20210917220251_remove_columns_from_users.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+class RemoveColumnsFromUsers < ActiveRecord::Migration[5.2]
+  def change
+    remove_column :users, :first_name, :string
+    remove_column :users, :last_name, :string
+    remove_column :users, :google_token, :string
```
Comment (carolinectan) [inline db/migrate/20210917220251_remove_columns_from_users.rb:5]: Thanks for updating this!

File: spec/factories/events.rb [hunk 7]
```diff
@@ -1,8 +1,14 @@
 FactoryBot.define do
   factory :event do
-    date_time { DateTime.now }
+    date_time { DateTime.now + rand(10).days }
```
Comment (carolinectan) [inline spec/factories/events.rb:3]: Cool solution for making events in the future!

File: spec/factories/events.rb [hunk 8]
```diff
@@ -1,8 +1,14 @@
...
     activity { Faker::Hobby.activity }
     user
     gym
   end
 end
+
+def event_with_invitation
```
Comment (carolinectan) [inline spec/factories/events.rb:10]: Can an event be created with an invitation?

File: app/controllers/api/v1/users_controller.rb (hunk 3 above)
Comment (tvaroglu) [reply app/controllers/api/v1/users_controller.rb:26]: If we can get the `get_user` action from FE working smoothly and efficiently, definitely.  Right now, the only time these get updated is upon initial user registration (we just direct straight to dashboard if login is successful, with the way current session controller flow is built)

File: app/serializers/user_serializer.rb [hunk 9]
```diff
@@ -1,11 +1,20 @@
...
   include JSONAPI::Serializer
 
-  attributes :email, :google_id, :google_token, :google_image_url, :zip_code,
-             :summary, :goal, :availability_morning, :availability_afternoon,
-             :availability_evening, :first_name, :last_name, :full_name
+  attributes :email, :google_id, :google_image_url, :zip_code, :summary, :goal,
+             :availability_morning, :availability_afternoon,
+             :availability_evening, :full_name
 
   # TODO: Add relationships so they can be included in serialized hash
```
Comment (tvaroglu) [inline app/serializers/user_serializer.rb:8]: The use of metadata on this will be an awesome solution.  Nice working ahead on this

File: db/migrate/20210917220251_remove_columns_from_users.rb (hunk 6 above)
Comment (tvaroglu) [reply db/migrate/20210917220251_remove_columns_from_users.rb:5]: ^^

File: spec/factories/users.rb (hunk 4 above)
Comment (tvaroglu) [reply spec/factories/users.rb]: 21 (according to mine)

File: spec/models/event_spec.rb [hunk 10]
```diff
@@ -19,11 +19,22 @@
...
       end
     end
+
+    describe '#event_with_invitation' do
+      subject(:event_with_invite) { event_with_invitation }
+
+      it 'is valid with valid attributes' do
+        expect(event_with_invite).to be_valid
+        expect(event_with_invite.invitations.size).to eq(1)
+        expect(event_with_invite.invitations.first).to be_an Invitation
```
Comment (carolinectan) [inline spec/models/event_spec.rb:35]: Very thorough testing!

File: spec/requests/api/v1/users/create_request_spec.rb [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  describe 'POST /api/v1/users' do
+    let(:valid_attributes) do
+      {
+        email: Faker::Internet.email,
+        google_id: Faker::Alphanumeric.alphanumeric(number: 10),
+        google_image_url: Faker::Avatar.image,
+        zip_code: Faker::Address.zip_code.first(5),
+        summary: Faker::GreekPhilosophers.quote,
+        goal: rand(0..5),
+        availability_morning: [true, false].sample,
```
Comment (carolinectan) [inline spec/requests/api/v1/users/create_request_spec.rb:14]: Good use of `.sample` instead of just setting all values to either true or false

File: spec/requests/api/v1/users/create_request_spec.rb [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        ["Email can't be blank",
+         "Google can't be blank",
+         "Google image url can't be blank",
+         "Zip code can't be blank",
+         "Zip code is not a number",
+         "Summary can't be blank",
+         "Full name can't be blank"]
+      end
+
+      before { post '/api/v1/users', params: { goal: 'value1' } }
```
Comment (carolinectan) [inline spec/requests/api/v1/users/create_request_spec.rb:54]: I haven't seen one-line syntax for a before :each. Neat!

File: app/controllers/api/v1/users_controller.rb (hunk 1 above)
Comment (Scott-Borecki) [reply app/controllers/api/v1/users_controller.rb:29]: Good point!  We'll have to coordinate that with FE peoples.

File: app/controllers/api/v1/users_controller.rb (hunk 2 above)
Comment (Scott-Borecki) [reply app/controllers/api/v1/users_controller.rb:34]: Good point!  We'll have to coordinate that with FE peoples.

File: spec/factories/users.rb (hunk 4 above)
Comment (Scott-Borecki) [reply spec/factories/users.rb]: Awesome -- thank you!  I updated this in the latest commits.

File: db/migrate/20210917220251_remove_columns_from_users.rb (hunk 6 above)
Comment (Scott-Borecki) [reply db/migrate/20210917220251_remove_columns_from_users.rb:5]: [insert party cat emoji]

File: spec/requests/api/v1/users/gyms/index_request_spec.rb [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    context 'when the user record does not exist' do
+      before { get "/api/v1/users/#{bad_user_id}/gyms" }
+
+      it 'returns an error message', :aggregate_failures do
+        expect(json).not_to be_empty
+        expect(json.size).to eq(2)
+
+        expect(json[:message]).to eq('your query could not be completed')
+
+        expect(json[:errors]).to be_an Array
```
Comment (carolinectan) [inline spec/requests/api/v1/users/gyms/index_request_spec.rb:38]: Consider consistency with using or not using parenthesis in your expected return value.

File: spec/factories/events.rb (hunk 8 above)
Comment (Scott-Borecki) [reply spec/factories/events.rb:10]: I think this is something we should flag for later.  Perhaps we include a validation on either the FE or BE.

**Review by tvaroglu — APPROVED** [review]
Amazing work!  Thank you for providing comments for helpers, this is fantastically collaborative groundwork that anyone on the team could pick up and run with.

File: spec/requests/api/v1/users/gyms/index_request_spec.rb (hunk 13 above)
Comment (Scott-Borecki) [reply spec/requests/api/v1/users/gyms/index_request_spec.rb:38]: Good catch!  Usually, I'll leave the parenthesis out for checking Class types.

File: spec/requests/api/v1/users/update_request_spec.rb [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+        before { patch "/api/v1/users/#{user.id}", params: valid_attributes }
+
+        it 'updates the user', :aggregate_failures do
+          user.reload
+
+          expect(json).not_to be_empty
+          expect(json_data.size).to eq(3)
+          expect(json_data[:attributes].size).to eq(10)
+          expect(json_data[:attributes][:summary]).to eq(summary)
```
Comment (carolinectan) [inline spec/requests/api/v1/users/update_request_spec.rb:40]: Maybe add that the values are the type expected?

Comment (Scott-Borecki) [reply spec/requests/api/v1/users/update_request_spec.rb:40]: Good catch!  This should get tested in the Serializer tests.

**Review by carolinectan — APPROVED** [review]
Very clean and concise code. Good use of single line syntax and support folder to dry up tests.
